server/automute_test.go

```diff
@@ -0,0 +1,451 @@
+	}
+	return channels, nil
+}
+
+func (a *AutomuteAPIMock) GetChannelMember(channelID, userID string) (*model.ChannelMember, *model.AppError) {
+	member, ok := a.channelMembers[a.key(channelID, userID)]
+	if !ok {
+		return nil, &model.AppError{Message: "Channel member not found"}
+	}
+	return member, nil
+}
+
+func (a *AutomuteAPIMock) PatchChannelMembersNotifications(identifiers []*model.ChannelMemberIdentifier, notifyProps map[string]string) *model.AppError {
+	for _, identifier := range identifiers {
+		for propKey, propValue := range notifyProps {
+			a.channelMembers[a.key(identifier.ChannelId, identifier.UserId)].NotifyProps[propKey] = propValue
+		}
+	}
+
+	return nil
+}
+
+func (a *AutomuteAPIMock) LogDebug(msg string, keyValuePairs ...any) {
+	a.t.Log(msg, keyValuePairs)
+}
+
+func (a *AutomuteAPIMock) LogInfo(msg string, keyValuePairs ...any) {
+	a.t.Log(msg, keyValuePairs)
+}
+
+func (a *AutomuteAPIMock) LogError(msg string, keyValuePairs ...any) {
+	a.t.Error(msg, keyValuePairs)
+}
+
+func (a *AutomuteAPIMock) LogWarn(msg string, keyValuePairs ...any) {
+	a.t.Log(msg, keyValuePairs)
+}
+
+func newAutomuteTestPlugin(t *testing.T) *Plugin {
+	mockAPI := &AutomuteAPIMock{
+		API:            &plugintest.API{},
+		channels:       make(map[string]*model.Channel),
+		channelMembers: make(map[string]*model.ChannelMember),
+		preferences:    make(map[string]model.Preference),
+		t:              t,
+	}
+	t.Cleanup(func() { mockAPI.AssertExpectations(t) })
+
+	p := newTestPlugin(t)
+	p.SetAPI(mockAPI)
+
+	return p
+}
+
+func TestShouldEnableAutomuteForUser(t *testing.T) {
+	user := &model.User{Id: model.NewId()}
+
+	t.Run("should return true when a user is both connected and has their primary platform set to Teams", func(t *testing.T) {
+		p := newAutomuteTestPlugin(t)
+
+		mockUserConnected(p, user.Id)
+
+		setPrimaryPlatform(p, user.Id, PreferenceValuePlatformMSTeams)
+
+		result, err := p.shouldEnableAutomuteForUser(user.Id, false, false)
+
+		assert.Equal(t, true, result)
+		assert.NoError(t, err)
+	})
+
+	t.Run("should return false when a user is not connected but has their primary platform set to Teams", func(t *testing.T) {
+		p := newAutomuteTestPlugin(t)
+
+		mockUserNotConnected(p, user.Id)
+
+		setPrimaryPlatform(p, user.Id, PreferenceValuePlatformMSTeams)
+
+		result, err := p.shouldEnableAutomuteForUser(user.Id, false, false)
+
+		assert.Equal(t, false, result)
+		assert.NoError(t, err)
+	})
+
+	t.Run("should return false when a user is connected but has their primary platform set to MM", func(t *testing.T) {
+		p := newAutomuteTestPlugin(t)
+
+		mockUserConnected(p, user.Id)
+
+		setPrimaryPlatform(p, user.Id, PreferenceValuePlatformMM)
+
+		result, err := p.shouldEnableAutomuteForUser(user.Id, false, false)
+
+		assert.Equal(t, false, result)
+		assert.NoError(t, err)
+	})
+
+	t.Run("should return false when a user is not connected and has their primary platform set to MM", func(t *testing.T) {
```

Comment:
Maybe not needed, but we could test in case no preference is set.

Reply:
Maybe I'm not treating this as enough of a black box, but I figured I'd handle that case when testing `isUsersPrimaryPlatformTeams` and then I could just assume that it works here. The other alternative would be to  rewrite `shouldEnableAutomute` so that it's called like 
```go
shouldEnableAutomute(p.isUserConnected(userID), p.isUsersPrimaryPlatformTeams(userID))
```
but then testing this function would be just testing that `return !connected || !teamsPrimary` returned true which is as pointless as it sounds.

Maybe that would be a better way to rewrite this, but I'm not going to rewrite it now